### ISSUE TITLE
How to adapt learning rate based on custom rules during training?

### ISSUE THREAD

**Issue by huangmozhilv** [tensorpack/tensorpack#1000]
+ If you did something with tensorpack and it failed, please use the "Unexpected Problems /
  Bugs" category.

+ Your question is probably answered in [tutorials](http://tensorpack.readthedocs.io/en/latest/tutorial/index.html#user-tutorials). Read it first.

+ We answer "HOW to do X with Tensorpack" for a well-defined X.
  We also answer "HOW/WHY Tensorpack does X" for some X that Tensorpack or its examples are doing.

Some typical questions that we DO NOT answer:

+ "Could you improve/implement an example/paper ?" --
  We have no plans to do so. We don't consider feature
  requests for examples or implement a paper for you.
  If you don't know how to do something yourself, you may ask a usage question.
+ "The examples do not perform well after I change the models/dataset/parameters/etc."
  Tensorpack maintainers make sure the examples perform well without modification.
  But it's your job to pick the model and parameters that are suitable for your own situation.
  We cannot help with such questions unless they appear to be a bug in tensorpack.
+ "Why my model doesn't work?", "I don't understand this paper you implement.",
  "How should I change the examples for my own dataset?"
  We do not answer machine learning questions. 

You can also use gitter (https://gitter.im/tensorpack/users) for more casual discussions.

The code is derived from `train.py` of [tkuanlun350/3DUnet-Tensorflow-Brats18](https://github.com/tkuanlun350/3DUnet-Tensorflow-Brats18), the lines I added are line 6, 7, 13, 14, 52-58. My learning rate adaption strategy is: compute the exponential moving average (EMA) of total cost, then reduce learning rate by factor 5 once EMA does not improve by at least 5e-3. However, with this code, the learning rate never changes.
```
1. class Unet3dModel(ModelDesc):
2.     def __init__(self, modelType="training", inference_shape=config.INFERENCE_PATCH_SIZE):
3.         self.modelType = modelType
4.         self.inference_shape = inference_shape
5.         print(self.modelType)
6.         self.lr = tf.get_variable('learning_rate', initializer=config.BASE_LR, trainable=False)
7.         self.last_total_cost_ema = tf.get_variable('last_total_cost_ema', initializer=1.0, trainable=False)
8. 
9.     def optimizer(self):
10.         # lr = tf.get_variable('learning_rate', initializer=config.BASE_LR, trainable=False)
11.         # tf.summary.scalar('learning_rate', lr)
12.         # opt = tf.train.MomentumOptimizer(lr, 0.9)
13.         tf.summary.scalar('learning_rate', self.lr)
14.         opt = tf.train.MomentumOptimizer(self.lr, 0.9)
15.         return opt
16.         
17.     def preprocess(self, image):
18.         # transform to NCHW
19.         return tf.transpose(image, [0, 4, 1, 2, 3])
20. 
21.     def inputs(self):
22.         S = config.PATCH_SIZE
23.         if self.modelType == 'training':
24.             ret = [
25.                 tf.placeholder(tf.float32, (config.BATCH_SIZE, S[0], S[1], S[2], 4), 'image'),
26.                 tf.placeholder(tf.float32, (config.BATCH_SIZE, S[0], S[1], S[2], 1), 'weight'),
27.                 tf.placeholder(tf.float32, (config.BATCH_SIZE, S[0], S[1], S[2], 1), 'label')]
28.         else:
29.             S = self.inference_shape
30.             ret = [
31.                 tf.placeholder(tf.float32, (config.BATCH_SIZE, S[0], S[1], S[2], 4), 'image')]
32.         return ret
33. 
34.     def build_graph(self, *inputs):
35.         is_training = get_current_tower_context().is_training
36.         if is_training:
37.             image, weight, label = inputs
38.         else:
39.             image = inputs[0]
40.         image = self.preprocess(image)
41.         featuremap = unet3d('unet3d', image) # final upsampled feturemap
42.         if is_training:
43.             loss = Loss(featuremap, weight, label)
44.             wd_cost = regularize_cost(
45.                     '(?:unet3d)/.*kernel',
46.                     l2_regularizer(1e-5), name='wd_cost')
47. 
48.             total_cost = tf.add_n([loss, wd_cost], 'total_cost')
49. 
50.             add_moving_summary(total_cost, wd_cost)
51. 
52.             # keep an exponential moving average
53.             ema = tf.train.ExponentialMovingAverage(decay=1-1/30)
54.             ema.apply([total_cost])
55.             ave = ema.average(total_cost)
56.             self.lr = tf.cond(tf.less(self.last_total_cost_ema-ave, 5e-3), lambda: self.lr/5, lambda:self.lr)
57.             # self.lr /= 5
58.             self.last_total_cost_ema = ave
59.             
60.             return total_cost
61.         else:
62.             final_probs = tf.nn.softmax(featuremap, name="final_probs") #[b,d,h,w,num_class]
63.             final_pred = tf.argmax(final_probs, axis=-1, name="final_pred")
```
